frontend/src/plugins/impl/plotly/PlotlyPlugin.tsx

```diff
@@ -109,6 +109,15 @@ export const PlotlyComponent = memo(
       return structuredClone(originalFigure);
     });
 
+    // Used for rendering LaTeX. TODO: Serve this library from Marimo
+    const mathJaxStatus = useScript(
+      "https://cdn.jsdelivr.net/npm/mathjax@2/MathJax.js?config=TeX-MML-AM_CHTML",
+      {
+        removeOnUnmount: true,
```

Comment:
The issue is that status will remain at "loading" for components rendered at the same time, 

Component 1 intialize -> calls useScript -> `status` becomes loading
Component 2 initialize -> calls useScript -> sees that `status` is loading (checks script status tag) -> doesn't do anything

`status` changes to ready -> triggers Component 1 re-render
Component 2 isn't re-rendered. Following your example, it would never be displayed.
[uidotdev/usehooks/useScript](https://github.com/uidotdev/usehooks/blob/90fbbb4cc085e74e50c36a62a5759a40c62bb98e/index.js#L1104C1-L1109C1)

Reply:
oh interesting, im going to see if i can refactor this to a singleton with jotai and give you another hook

Reply:
Maybe this would help:

```typescript
import { atom, selectAtom, useAtomValue, useSetAtom } from 'jotai';
import { useEffect, useRef } from 'react';

// Types
type ScriptStatus = 'loading' | 'ready' | 'error' | 'unknown';
type ScriptStatuses = Record<string, ScriptStatus>;
type ScriptOptions = {
  removeOnUnmount?: boolean;
};

// Atoms
const scriptStatusesAtom = atom<ScriptStatuses>({});

// Create a selector factory for individual script status
const createScriptStatusSelector = (src: string) => 
  selectAtom(
    scriptStatusesAtom,
    (statuses) => statuses[src] || 'loading'
  );

// Helper function to manage script loading
const loadScript = (
  src: string, 
  script: HTMLScriptElement, 
  setStatus: (status: ScriptStatus) => void
): (() => void) => {
  script.setAttribute("data-status", "loading");

  const handleScriptLoad = () => {
    script.setAttribute("data-status", "ready");
    setStatus("ready");
    removeEventListeners();
  };

  const handleScriptError = () => {
    script.setAttribute("data-status", "error");
    setStatus("error");
    removeEventListeners();
  };

  const removeEventListeners = () => {
    script.removeEventListener("load", handleScriptLoad);
    script.removeEventListener("error", handleScriptError);
  };

  script.addEventListener("load", handleScriptLoad);
  script.addEventListener("error", handleScriptError);

  return removeEventListeners;
};

// Memoized selector cache
const scriptStatusSelectors: Record<string, ReturnType<typeof createScriptStatusSelector>> = {};

export function useScript(src: string, options: ScriptOptions = {}): ScriptStatus {
  // Get or create a memoized selector for this src
  if (!scriptStatusSelectors[src]) {
    scriptStatusSelectors[src] = createScriptStatusSelector(src);
  }
  
  const status = useAtomValue(scriptStatusSelectors[src]);
  const setScriptStatuses = useSetAtom(scriptStatusesAtom);
  const optionsRef = useRef(options);

  useEffect(() => {
    let script = document.querySelector(`script[src="${src}"]`);

    const updateStatus = (newStatus: ScriptStatus) => {
      setScriptStatuses(prev => ({
        ...prev,
        [src]: newStatus
      }));
    };

    const domStatus = script?.getAttribute("data-status") as ScriptStatus | null;
    if (domStatus) {
      updateStatus(domStatus);
      return;
    }

    if (script === null) {
      script = document.createElement("script");
      script.src = src;
      script.async = true;
      
      document.body.appendChild(script);
      updateStatus("loading");

      const removeEventListeners = loadScript(src, script, updateStatus);

      const removeOnUnmount = optionsRef.current.removeOnUnmount;

      return () => {
        if (removeOnUnmount === true) {
          script.remove();
          removeEventListeners();
          setScriptStatuses(prev => {
            const newStatuses = { ...prev };
            delete newStatuses[src];
            return newStatuses;
          });
        }
      };
    } else {
      updateStatus("unknown");
    }
  }, [src, setScriptStatuses]);

  return status;
}

```

Reply:
interesting, let me have a look in the morning haha.

Reply:
i'm wondering if using an observer on the DOM would be more lightweight / easier to reason about. Using jotai would be tracking multiple script statuses

```.tsx
export function useScriptCustom(
  src: string,
  options: ScriptOptions = defaultOptions,
) {
  const [status, setStatus] = React.useState<ScriptStatus>();

  React.useEffect(() => {
    let script: HTMLScriptElement | null = document.querySelector(
      `script[src="${src}"]`,
    );

    const updateStatus = (newStatus: ScriptStatus) => {
      if (script) {
        script.dataset.status = newStatus;
        setStatus(newStatus);
      }
    };

    const handleScriptLoad = () => updateStatus("ready");
    const handleScriptError = () => updateStatus("error");

    // observes DOM changes and updates the status
    const observeScript = (script: HTMLScriptElement) => {
      const observer = new MutationObserver(() => {
        const newStatus = script.dataset.status;
        if (newStatus) {
          setStatus(newStatus as ScriptStatus);
        }
      });
      observer.observe(script, {
        attributes: true,
        attributeFilter: ["data-status"],
      });
      return observer;
    };

    if (script) {
      const domStatus = script.dataset.status;
      if (domStatus) {
        setStatus(domStatus as ScriptStatus);
        const observer = observeScript(script);
        return () => observer.disconnect();
      }
    } else {
      script = document.createElement("script");
      script.src = src;
      script.async = true;
      updateStatus("loading");
      document.body.append(script);

      script.addEventListener("load", handleScriptLoad);
      script.addEventListener("error", handleScriptError);

      const observer = observeScript(script);

      return () => {
        observer.disconnect();
        if (script) {
          script.removeEventListener("load", handleScriptLoad);
          script.removeEventListener("error", handleScriptError);

          if (options.removeOnUnmount) {
            script.remove();
          }
        }
      };
    }
    setStatus("unknown");
  // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [src]);

  return status;
}
```

Reply:
@Light2Dark - if the observer works, open to use that instead!